### DISCUSSION TITLE
[Typescript] Component does not have any construct or call signatures

### DISCUSSION THREAD

**Discussion by loganaaaflag** [rsuite/rsuite#1620] [Q&A]
Hi!

I've just started using rsuite - it's been really amazing so far. However, I'm running into various type issues - one of which is that "sub components" attached as properties to a function component are marked as optional and thus giving me the following error:

![image](https://user-images.githubusercontent.com/55156743/115073494-44754900-9ead-11eb-9a16-d9b1beeaf472.png)

I'm guessing the properties were marked optional in the TypeScript interfaces to allow adding them after declaring the main component (like the [modal example here](https://github.com/rsuite/rsuite/blob/7e910df5f610b96af41de9b1a5bc8427feb531c3/src/Modal/Modal.tsx#L205))

What strategies are people using to cleanly resolve these type issues?

**Comment by loganaaaflag** [comment]
It seems like it should be resolved in the code base itself by removing the optional flag in the interface and opting for a strategy like 

```ts
type ModalBaseComponent = RsRefForwardingComponent<'div', ModalProps>;

type ModalComponent = ModalBaseComponent & {
    Body: typeof ModalBody;                                                                                                                                                                      
    Header: typeof ModalHeader;     
    // ...
}

const ModalBase: ModalBaseComponent = React.forwardRef((props, ref) => {
    // ...
});

const Modal: ModalComponent = Object.assign(ModalBase, {
    Header: ModalHeader,
    Footer: ModalFooter,
    // ...
});

export default Modal;
```

Or all the child components could consistently be exported separately like how Form child components are. 

Thoughts?

**Comment by loganaaaflag** [comment]
I'm currently importing the modules separately.

```tsx
import Modal from 'rsuite'
import ModalBody from "rsuite/es/Modal/ModalBody";
import ModalHeader from "rsuite/es/Modal/ModalHeader";   

return (
    <Modal>
        <ModalHeader>...</ModalHeader>
    </Modal>`
)
```

**Comment by simonguo** [comment]
It has been fixed https://github.com/rsuite/rsuite/pull/1640  

```
npm i rsuite@next
```
